### ISSUE TITLE
New feature request

### ISSUE THREAD

**Issue by lxhuang7** [hartleys/QoRTs#5]
Hi Stephen,

Thanks for this wonderful tool. I was wondering if you consider to add a feature for estimating the number of reads originated from ribosomal RNAs. This is important because many RNA libraries, e.g. FFPE samples, contain relatively large amount of rRNAs even after the ribo-depletion protocols are applied.

Best regards,
Lei

**Comment by hartleys** [comment]
Shouldn't be a problem. I'll probably have a new version with this functionality released by Monday.

What I'm actually going to do is generate counts for each "BioType" using the BioType tag used by Ensembl GTF files. Unfortunately there's no universal specification for how to mark rRNAs. Most databases seems to use their own style. I can't think of a straightforward way to support all possible formats: there are just so many and they're changing all the time.

I'm going to try to set it up so that it's fairly easy to add the information to the gtf.

**Comment by lxhuang7** [comment]
Great! Thanks! Many people use rRNA annotation from RepeatMasker database, which has a few more entries than that in Gencode or Ensembl gtf file.

**Comment by hartleys** [comment]
Can you provide a link to exactly which file you mean?

**Comment by lxhuang7** [comment]
One way to retrieve rRNA annotation is to get it from UCSC table browser (see this tip at http://goo.gl/CjsnZB). The annotation file is actually from RepeatMasker. RSeQC website contains a rRNA bed file for human,  (http://goo.gl/hOy4Dd) but it is for hg19 and outdated.

**Comment by hartleys** [comment]
I added a biotype-counting plot. It can be accessed via the "makePlot.biotype.rates" function. See the [FAQ](http://hartleys.github.io/QoRTs/FAQ.html) for documentation on how exactly it works and what you will need to do.

A few caveats about using QoRTs to estimate rRNA quantities:

rRNA's often have a TON of copies across the genome, so you would probably get massive multi-mapping for these features. QoRTs currently ignores multi-mapped reads, so the counts might be much lower than expected. In addition, even if I code QoRTs to count multi-mapped reads for this purpose, many aligners will refuse to map a read over too many different loci and just dump the read as unmappable. 

As a result, heightened rRNA quantities might cause a noticeable uptick in the multi-mapping rate and a downtick in the mapping rate. You might even be able to infer rRNA quantity outliers just from that. But really, in a perfect world, you'd want to use some other method to estimate rRNA quantities from raw pre-mapped reads. Not sure how you'd go about that.

I don't have any FFPE data on-hand to test this stuff on. I can probably find some, but that'll take a while. So all this stuff is mostly just educated guesswork. Your results may vary.

For future updates I intend to make this function a bit more flexible and implement support for other input data formats. Right now it requires a GTF file which includes "gene_biotype" tags for the desired genes. Note that you do NOT need to add this tag to every line or even every gene. Just the genes you want counted. Any genes not assigned a biotype will be counted as biotype "UNK".

I've attached an example of what the biotype plot looks like on the QoRTs example dataset. Note the log scale. Also note that there are VERY few rRNA reads. The example data is from an extremely high quality ribo-zero gold dataset, so this isn't too terribly surprising, particularly given the caveats noted above.

![testbiotype](https://cloud.githubusercontent.com/assets/2895596/11042779/18f318d6-86e7-11e5-92c4-5f34b9fa690b.png)

**Comment by lxhuang7** [comment]
Thanks a lot! Up to now I use Picard CollectRnaSeqMetrics tool to count the number and proportion of reads mapped to rRNA. Beside rRNA, the tool provides the percentage of reads mapped to mRNA (coding and non-coding), 3'UTR, 5'UTR, intergenic and intragenic region. It might be an alternative to biotype grouping.

**Comment by hartleys** [comment]
The "Gene Assignment Rates" plot (see section 7.4.14 in the vignette) already covers most of the functionality you describe. It includes counts for "Unique Gene" (which means the read maps to one gene's exons, CDS or UTR), "UTR", intronic, 1kb from nearest gene, 10kb from nearest gene, or >10kb from nearest gene ("middle of nowhere").

Note that picard CollectRnaSeqMetrics takes the same basic information used by QoRTs, and is thus still subject to the same limitations and caveats I previously described: reads map poorly to the rRNA regions and thus your rRNA counts may be deceptively low.

Bottom line: Measuring rRNA might be tricky. Do be careful.

**Comment by hartleys** [comment]
You can also find the counts themselves in the summary table produced by makeMultiPlot.all() or get.summary.table().

The "FIELD" columns are:

    ReadPairs_AmbigGene
    ReadPairs_UniqueGene
    ReadPairs_UniqueGene_CDS	
    ReadPairs_UniqueGene_UTR	
    ReadPairs_NoGene
    ReadPairs_NoGene_Intron
    ReadPairs_NoGene_OneKbFromGene	
    ReadPairs_NoGene_TenKbFromGene	
    ReadPairs_NoGene_MiddleOfNowhere

**Comment by lxhuang7** [comment]
Thanks for clarification. Thanks again for this awesome tool!
